[LoongArch] Emit R_LARCH_RELAX when expanding some LoadAddress

MQ-mengqing opened the pull request:
Emit relax relocs when expand non-large la.pcrel and non-large la.got on llvm-mc stage, which like what does on GAS.
1, la.pcrel -> PCALA_HI20 + RELAX + PCALA_LO12 + RELAX
2, la.got -> GOT_PC_HI20 + RELAX + GOT_PC_LO12 + RELAX

MQ-mengqing committed:
Emit relax relocs when expand some LoadAddress

Emit relax relocs when expand non-large la.pcrel and non-large la.got
on llvm-mc stage, which like what does on GAS.
1, la.pcrel -> PCALA_HI20 + RELAX + PCALA_LO12 + RELAX
2, la.got -> GOT_PC_HI20 + RELAX + GOT_PC_LO12 + RELAX

MQ-mengqing commented:
Add: @SixWeining @wangleiat @xry111 @xen0n @MaskRay .

MaskRay commented on llvm/test/MC/LoongArch/Relocations/relax-la.s:
```diff
@@ -0,0 +1,19 @@
+.text
+.L1:
+  la.pcrel $a0, .L1
+  la.got   $a0, .L1
```
`llvm/test/MC/LoongArch/Macros/macros-la.s` can be enhanced to test relocations.

MaskRay commented:
> Emit relax relocs when expand some LoadAddress

I suggest `[LoongArch] Emit R_LARCH_RELAX when expanding some LoadAddress`

MQ-mengqing committed:
Delete relax-la test and test it in macros-la

SixWeining commented on llvm/test/MC/LoongArch/Macros/macros-la.s:
```diff
@@ -1,66 +1,110 @@
 la.pcrel $a0, sym_pcrel
 # CHECK:      pcalau12i $a0, %pc_hi20(sym_pcrel)
 # CHECK-NEXT: addi.d $a0, $a0, %pc_lo12(sym_pcrel)
+# RELOC:      R_LARCH_PCALA_HI20 sym_pcrel 0x0
```
Had better to use `RELOC-NEXT` for more accurate.

Seems we also should change `CHECK` to `CHECK-NEXT` before this PR.

SixWeining commented on llvm/test/MC/LoongArch/Macros/macros-la.s:
```diff
@@ -1,66 +1,110 @@
 # CHECK-NEXT: ori $a0, $a0, %abs_lo12(sym_abs)
 # CHECK-NEXT: lu32i.d $a0, %abs64_lo20(sym_abs)
 # CHECK-NEXT: lu52i.d $a0, $a0, %abs64_hi12(sym_abs)
+# RELOC:      R_LARCH_ABS_HI20 sym_abs 0x0
```
It's better to add start/end flags like: 
```
# RELOC:          Relocations [
# RELOC-NEXT:       Section ({{.*}}) .rela.text {

... your check goes here

# RELOC-NEXT:       }
# RELOC-NEXT:     ]
```

MQ-mengqing committed:
Update tests due to #72960 merged and address @SixWeining's comments.

1, update relax-addsub test to check `la.pcrel` RELAX.
2, update subsection test to check RELAX.
3, add check reloc start/end flags.
4, change non-first CHECK to CHECK-NEXT and add CHECK-EMPTY.

SixWeining reviewed (approved):
LGTM
